### PR TITLE
Add broadcast event api

### PR DESCRIPTION
**Pull request by danvixent**
Add support for broadcasting events to all endpoints while respecting event_type filter

### CODE REVIEW

File: api/api.go [hunk 1]
```diff
@@ -142,6 +142,7 @@ func (a *ApplicationHandler) BuildRoutes() *chi.Mux {
 						// TODO(all): should the InstrumentPath change?
 						eventRouter.With(middleware.InstrumentPath("/events")).Post("/", handler.CreateEndpointEvent)
 						eventRouter.Post("/fanout", handler.CreateEndpointFanoutEvent)
+						eventRouter.Post("/broadcast", handler.CreateBraodcastEvent)
```
Comment (subomi) [inline api/api.go:145]: @danvixent, I think there's a typo here. The handler name should be `CreateBroadcastEvent`.

File: worker/task/process_broadcast_event_creation.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+		if len(broadcastEvent.IdempotencyKey) > 0 {
+			events, err := eventRepo.FindEventsByIdempotencyKey(ctx, broadcastEvent.ProjectID, broadcastEvent.IdempotencyKey)
+			if err != nil {
+				return &EndpointError{Err: err, delay: 10 * time.Second}
+			}
+
+			isDuplicate = len(events) > 0
+		}
+
+		subscriptions, err := subRepo.FindSubscriptionByEventType(ctx, project.UID, broadcastEvent.EventType)
```
Comment (subomi) [inline worker/task/process_broadcast_event_creation.go:47]: @danvixent We should retrieve all subscriptions in the project and run both `matchSubscriptions` and `matchSubscriptionsUsingFilter` routines in `process_event_creation.go` against them.

**Review by subomi — CHANGES_REQUESTED** [review]
@danvixent FYI

File: worker/task/process_broadcast_event_creation.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+func ProcessBroadcastEventCreation(endpointRepo datastore.EndpointRepository, eventRepo datastore.EventRepository, projectRepo datastore.ProjectRepository, eventDeliveryRepo datastore.EventDeliveryRepository, eventQueue queue.Queuer, subRepo datastore.SubscriptionRepository, deviceRepo datastore.DeviceRepository) func(context.Context, *asynq.Task) error {
+	return func(ctx context.Context, t *asynq.Task) error {
+		var broadcastEvent models.BroadcastEvent
+
+		err := msgpack.DecodeMsgPack(t.Payload(), &broadcastEvent)
+		if err != nil {
+			err := json.Unmarshal(t.Payload(), &broadcastEvent)
+			if err != nil {
+				return &EndpointError{Err: err, delay: defaultDelay}
+			}
```
Comment (jirevwe) [inline worker/task/process_broadcast_event_creation.go:30]: Do we need this since we're sure it will be msgpack?

**Review by jirevwe — APPROVED** [review]

File: worker/task/process_broadcast_event_creation.go (hunk 3 above)
Comment (danvixent) [reply worker/task/process_broadcast_event_creation.go:30]: @jirevwe how do you mean we're sure?

Comment (danvixent) [reply worker/task/process_broadcast_event_creation.go]: @jirevwe thanks, fixed!

**Review by jirevwe — APPROVED** [review]

**Pull request merged**